File: js/cryptus/aes.js
const helpers = require("./helpers");

async function importAesKeyFromHex(passwordHex) {
  const crypto = helpers.getCrypto();
  const keyBytes = helpers.hexToArrayBuffer(passwordHex);
  if (![16, 24, 32].includes(keyBytes.length)) {
    throw new Error("AES key must be 16/24/32 bytes");
  }
  return crypto.subtle.importKey("raw", keyBytes, { name: "AES-GCM" }, false, ["encrypt", "decrypt"]);
}

async function encryptAESGCMHex(plainText, passwordHex, nonceHex) {
  const crypto = helpers.getCrypto();
  const key = await importAesKeyFromHex(passwordHex);
  const iv = helpers.hexToArrayBuffer(nonceHex);
  if (iv.length !== 12) throw new Error("AES-GCM nonce must be 12 bytes");
  const ct = await crypto.subtle.encrypt({ name: "AES-GCM", iv }, key, helpers.textToArrayBuffer(plainText));
  return helpers.arrayBufferToHex(ct);
}

async function decryptAESGCMHex(cipherHex, passwordHex, nonceHex) {
  const crypto = helpers.getCrypto();
  const key = await importAesKeyFromHex(passwordHex);
  const iv = helpers.hexToArrayBuffer(nonceHex);
  if (iv.length !== 12) throw new Error("AES-GCM nonce must be 12 bytes");
  const ct = helpers.hexToArrayBuffer(cipherHex);
  const pt = await crypto.subtle.decrypt({ name: "AES-GCM", iv }, key, ct);
  return helpers.arrayBufferToText(pt);
}

module.exports = {
  encryptAESGCMHex,
  decryptAESGCMHex,
};

File: cryptus/aes.go
package cryptus

import (
	"crypto/aes"
	"crypto/cipher"
	"encoding/hex"
	"errors"
	"fmt"
)

const aesGcmNonceSize = 12 // default GCM

func (c *cryptus) EncryptAESGCMHex(plainText, keyHex, nonceHex string) (string, error) {
	key, err := hex.DecodeString(keyHex)
	if err != nil {
		return "", fmt.Errorf("invalid key hex: %w", err)
	}
	if l := len(key); l != 16 && l != 24 && l != 32 {
		return "", errors.New("AES key must be 16, 24, or 32 bytes")
	}
	nonce, err := hex.DecodeString(nonceHex)
	if err != nil {
		return "", fmt.Errorf("invalid nonce hex: %w", err)
	}
	if len(nonce) != aesGcmNonceSize {
		return "", fmt.Errorf("AES-GCM nonce must be %d bytes", aesGcmNonceSize)
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}
	gcm, err := cipher.NewGCMWithNonceSize(block, aesGcmNonceSize)
	if err != nil {
		return "", err
	}
	ct := gcm.Seal(nil, nonce, []byte(plainText), nil)
	return hex.EncodeToString(ct), nil
}

func (c *cryptus) DecryptAESGCMHex(cipherHex, keyHex, nonceHex string) (string, error) {
	key, err := hex.DecodeString(keyHex)
	if err != nil {
		return "", fmt.Errorf("invalid key hex: %w", err)
	}
	if l := len(key); l != 16 && l != 24 && l != 32 {
		return "", errors.New("AES key must be 16, 24, or 32 bytes")
	}
	nonce, err := hex.DecodeString(nonceHex)
	if err != nil {
		return "", fmt.Errorf("invalid nonce hex: %w", err)
	}
	if len(nonce) != aesGcmNonceSize {
		return "", fmt.Errorf("AES-GCM nonce must be %d bytes", aesGcmNonceSize)
	}
	raw, err := hex.DecodeString(cipherHex)
	if err != nil {
		return "", fmt.Errorf("invalid cipher hex: %w", err)
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}
	gcm, err := cipher.NewGCMWithNonceSize(block, aesGcmNonceSize)
	if err != nil {
		return "", err
	}
	pt, err := gcm.Open(nil, nonce, raw, nil)
	if err != nil {
		return "", err
	}
	return string(pt), nil
}
